### ISSUE TITLE
StreamingFileUpload subscriber "onComplete" not invoked sometimes

### ISSUE THREAD

**Issue by lubomir-moric** [micronaut-projects/micronaut-core#6532]
### Expected Behavior

- uploading file thru Controller
- subscribe to StreamingFileUpload
- when Publisher is done, onComplete is invoked on subscriber

### Actual Behaviour

- uploading file thru Controller
- subscribe to StreamingFileUpload
- onComplete() is never invoked

### Steps To Reproduce

Issue is happening randomly once in a while (e.g. every 20th REST call) and only! with file size of approx 32KB when first `onNext(pd: PartData) { ... }` receives already all data sent by client.

Here is a very simplified code in Kotlin I have:

Controller can accept 1 or more files (issues happening when 1 file is uploaded):
```
@Post("/upload")
@Consumes(MediaType.MULTIPART_FORM_DATA)
open fun upload(files: Publisher<StreamingFileUpload>): Mono<MutableHttpResponse<String>> {
  return Flux.from(files)
    .subscribeOn(Schedulers.boundedElastic())
    .flatMap { file ->
      Mono.create<String> { sink ->
        file.subscribe(MySubscriber(sink))
      }.map { "Uploaded FooBar" }
    }
    .reduce { _, _ -> "Upload summary FooBar }
    .map { HttpResponse.ok(it) }
}
```
Subscriber implementation:
```
class MySubscriber(private val sink: MonoSink<String>) : Subscriber<PartData> {
  private lateinit var subscription: Subscription
  private val stream: OutputStream = ByteArrayOutputStream()  // In reality I have Alex Mojaki's MultiPartOutputStream here, this is just for simplification

  override fun onSubscribe(s: Subscription) {
        subscription = s
        subscription.request(1)
  }

  override fun onNext(pd: PartData) {
    val bytes = pd.bytes
    stream.write(bytes)
    subscription.request(1)
  }

  override fun onError(t: Throwable) {
    sink.error(t)
  }

  override fun onComplete() {
    stream.close()
    sink.success("Finished")
  }
}
```

When added logging into `onNext` method on `PardData.bytes` size, when PartData size equals to whole upload file size, onComplete is not called (whole file content is sent within first `onNext`), HTTP client waits till timeout for any HTTP response, which is not arriving, as the Flow is not finishing.

Basically the code is the same since Micronaut 1.x.x, when migrated to Micronaut 3.0.x, later 3.1.x, this issue appeared.

Also I am not able to reproduce this behaviour with different file sizes (MB, GB ...), with smaller files the frequency of the issue is way less, like once in few hours, when doing serial continuous invocation of the API.

### Environment Information

Kotlin 1.5
JVM 11

### Version

3.1.4

**Comment by yawkat** [comment]
I can't reproduce this. I've added logging to `MySubscriber.onComplete` and `.onNext`, and the `onComplete` is always called for my `curl -v -F 'files=@input32k' http://localhost:8080/upload` command.

It's not too surprising that I can't reproduce it if the error happens as rarely as you say, there might be small changes in the code that cause the timing to change so that the issue doesn't happen anymore. But I can't debug this further without a test case, so please try to simplify your project where you see this issue to the point where you can share it.

**Comment by lubomir-moric** [comment]
Hello,
now I am able to reproduce this issue using a Spock test, the behaviour is exactly the same as it happens sometimes with curl, but for the test I construct the HTTP request manually to have a better control on the payload. With curl and even testing towards local application is hard to spot the issue, as the chunk onNext() is very fast and very small. Having the service on the network, it usually receives the whole 32KB chunk on first onNext() and halts.
With the following test, processing got stuck even on way smaller payloads, like 222B.

Even more simplified controller, now using built-in `transferTo(...)` method:
```
    @Post("/testupload2")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    open fun uploadTest2(dataFile: Flux<StreamingFileUpload>): Publisher<MutableHttpResponse<String>> {
        val os = BlackHoleOutputStream()
        println("uploadTest2")
        return dataFile
            .flatMap { x ->
                println("uploadTest2: flatmap")
                x.transferTo(os)
            }
            .map { success ->
                println("uploadTest2: map - $success")
                if (success)
                    ok("uploaded size: ${os.bytes}")
                else
                    status(HttpStatus.INTERNAL_SERVER_ERROR, "Something bad happened")
            }
            .onErrorReturn (
                status(HttpStatus.INTERNAL_SERVER_ERROR, "Something really bad happened")
            )
    }
```
```
class BlackHoleOutputStream : OutputStream() {
    var bytes: Long = 0

    override fun write(b: Int) {
        bytes++
        println("BlackHoleOutputStream: ${bytes}")
    }
}
```
Test client able to invoke the issue in Groovy / Spock; HTTP client never receives a response and `transferTo(...)` never completes:
```
import org.jetbrains.annotations.NotNull
import spock.lang.Specification

import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.ThreadLocalRandom
import java.util.function.Supplier

class HttpClientTest extends Specification {
    def boundaryString = '----*+*+*+*+*+*+*+*+*+*+'

    /**
     * Initial part of file upload field
     */
    InputStream uploadField(String name, String filename, String contentType) {
        def field = """\r
--${boundaryString}\r
Content-Disposition: form-data; name="${name}"; filename="${filename}"\r
Content-Type: ${contentType}\r
\r
"""
        new ByteArrayInputStream(field.bytes)
    }

    HttpRequest uploadRequest() {
        def url = "http://localhost:8080/test-api/testupload2"
        def sizeInBytes = 222
        def data = new byte[sizeInBytes]
        new Random().nextBytes(data)
        def inputStreams = [
                uploadField('dataFile', 'foo.rnd', 'application/octet-stream'),
                new ByteArrayInputStream(data),
                new ByteArrayInputStream("\r\n--${boundaryString}--\r\n".bytes)
        ] as List<InputStream>

        HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header('content-type', "multipart/form-data;boundary=${boundaryString}")
                .header('accept', 'application/json')
                .POST(HttpRequest.BodyPublishers.ofInputStream(new Supplier<InputStream>() {
                    @Override
                    InputStream get() {
                        new SequenceInputStream(Collections.enumeration(inputStreams))
                    }
                }))
                .build()
    }

    def uploadTest() {
        when:
        def client = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_1_1)
                .connectTimeout(Duration.ofSeconds(20))
                .build()

        def request = uploadRequest()
        def response = client.send(request, HttpResponse.BodyHandlers.ofString()) as HttpResponse<String>

        println 'status code: ' + response.statusCode()
        println response.body()

        then:
        1 == 1
    }
}
```
Note: when you increase the size of the file to e.g. ~2.22MB (2222222B), there is no issue at all.

**Comment by yawkat** [comment]
Thanks, I can reproduce it with your new code. Looks like an ugly race condition, will fix.
